packages/contracts/contracts/universal/TokenMultiSigWallet.sol

```diff
@@ -0,0 +1,238 @@
+     *
+     * @param _transactionId Index of submitted transaction requested.
+     */
+    modifier transactionExists(uint256 _transactionId) {
+        require(
+            transactions[_transactionId].target != address(0),
+            "TokenMultiSigWallet: transaction does not exist"
+        );
+        _;
+    }
+
+    /**
+     * @notice Ensure that the address is not zero address.
+     *
+     * @param _address Address resource requested.
+     */
+    modifier validAddress(address _address) {
+        require(_address != address(0), "TokenMultiSigWallet: address is not valid");
+        _;
+    }
+
+    /**
+     * @param _governor Address of the Governor contract.
+     */
+    constructor(address payable _governor) {
+        GOVERNOR = UpgradeGovernor(_governor);
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function submitTransaction(
+        address _target,
+        uint256 _value,
+        bytes memory _data
+    ) public onlyTokenOwner(msg.sender) returns (uint256) {
+        return _submitTransaction(_target, _value, _data);
+    }
+
+    function _submitTransaction(
+        address _target,
+        uint256 _value,
+        bytes memory _data
+    ) internal validAddress(_target) returns (uint256) {
+        uint256 transactionId = generateTransactionId(_target, _value, _data);
+        require(
+            transactions[transactionId].target == address(0),
+            "TokenMultiSigWallet: transaction already exists"
+        );
+
+        transactions[transactionId] = Types.MultiSigTransaction({
+            target: _target,
+            value: _value,
+            data: _data,
+            executed: false
+        });
+        emit TransactionSubmitted(msg.sender, transactionId);
+        return transactionId;
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function confirmTransaction(uint256 _transactionId)
+        public
+        onlyTokenOwner(msg.sender)
+        transactionExists(_transactionId)
+    {
+        Types.MultiSigConfirmation storage confirms = confirmations[_transactionId];
+        require(!confirms.confirmedBy[msg.sender], "TokenMultiSigWallet: already confirmed");
+        confirms.confirmedBy[msg.sender] = true;
+        confirms.confirmationCount += getVotes(msg.sender);
+        emit TransactionConfirmed(msg.sender, _transactionId);
+
+        // execute transaction if condition is met.
+        if (confirmations[_transactionId].confirmationCount >= quorum()) {
+            executeTransaction(_transactionId);
+        }
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function revokeConfirmation(uint256 _transactionId)
+        public
+        onlyTokenOwner(msg.sender)
+        transactionExists(_transactionId)
+    {
+        require(isConfirmedBy(_transactionId, msg.sender), "TokenMultiSigWallet: not confirmed yet");
+        require(!transactions[_transactionId].executed, "TokenMultiSigWallet: already executed");
```

Comment:
How about add `notExecuted` modifier for this function and `executeTransaction`?